core/src/main/kotlin/xtdb/compactor/SegmentMerge.kt

```diff
@@ -76,48 +78,120 @@ internal class SegmentMerge(private val al: BufferAllocator) : AutoCloseable {
+        val iidVec = dataReader["_iid"].rowCopier(relWriter["_iid"])
+        val sysFromVec= dataReader["_system_from"].rowCopier(relWriter["_system_from"])
+        val validFromVec = relWriter["_valid_from"]
+        val validToVec = relWriter["_valid_to"]
+        val opCopier = dataReader["op"].rowCopier(relWriter["op"])
+
+        val evPtr = EventRowPointer(dataReader, path)
+        val polygonCalculator = PolygonCalculator()
+
+        var seenErase = false
+
+        while (evPtr.isValid(isValidPtr, path)) {
+            evPtr.getIidPointer(startIidPtr)
+            curSystemFrom = evPtr.systemFrom
+
+            while(evPtr.isValid(isValidPtr, path) && evPtr.systemFrom == curSystemFrom && startIidPtr == evPtr.getIidPointer(curIidPtr)) {
+
+                when (val polygon = polygonCalculator.calculate(evPtr)) {
+                    // here we are only taking care of an erase that happens within the same transaction
+                    null -> {
+                        if (!seenErase) {
+                            iidVec.copyRow(evPtr.index)
+                            sysFromVec.copyRow(evPtr.index)
+                            validFromVec.writeLong(MIN_VALUE)
+                            validToVec.writeLong(MAX_VALUE)
+                            opCopier.copyRow(evPtr.index)
+                            relWriter.endRow()
+                        }
+                        seenErase = true
+                    }
+                    else -> {
+                        repeat(polygon.validTimeRangeCount) { i ->
+                            if (polygon.getSystemTo(i) > curSystemFrom) {
+                                iidVec.copyRow(evPtr.index)
+                                sysFromVec.copyRow(evPtr.index)
+                                validFromVec.writeLong(polygon.getValidFrom(i))
+                                validToVec.writeLong(polygon.getValidTo(i))
+                                opCopier.copyRow(evPtr.index)
+                                relWriter.endRow()
+                            }
+                        }
+                    }
+                }
+                evPtr.nextIndex()
+            }
+            polygonCalculator.reset()
+            seenErase = false
+        }
+        return relWriter
+    }
+
+    private fun MergePlanTask.merge(outWriter: OutWriter, pathFilter: ByteArray?, bitemporalResolution: Boolean = false) {
         val isValidPtr = ArrowBufPointer()
 
         val mpNodes = mpNodes
         val path = path.let { if (pathFilter == null || it.size > pathFilter.size) it else pathFilter }
         val mergeQueue = PriorityQueue(Comparator.comparing(QueueElem::evPtr, EventRowPointer.comparator()))
 
-        for (dataReader in mpNodes.mapNotNull { it.loadDataPage() }) {
-            val evPtr = EventRowPointer(dataReader, path)
-            val rowCopier = outWriter.rowCopier(dataReader)
+        val dataReaders = mutableListOf<RelationReader>()
```

Comment:
it seems that this `bitemporalResolution` flag is quite intertwined in the below logic, to the extent I found it relatively onerous to figure out which parts happen during which mode. did you consider duplicating it? if so, how'd it go?